### ISSUE TITLE
[Postgres] CDC replication mode to specify source defined cursor 

### ISSUE THREAD

**Issue by prateekmukhedkar** [airbytehq/airbyte#26807]
### What area the feature impact?

Connectors

### Revelant Information

The ask is for CDC enabled database sources to:
1. Specify `"source_defined_cursor": true` in the catalog 
2. Come up with a name for a cursor field and set `default_cursor_field` as this field name
3. Provide a value during sync for this cursor field - this value needs to be sortable within the scope of a primary key. It will be used by Normalization to dedup duplicate rows with the same primary key and chose the most recent row. 

There is no requirement for the cursor value to be sortable across primary keys, however if implementation makes it easier to generate cursor values with this property then it is ok. 

Relevant Airbyte protocol documentation https://docs.airbyte.com/understanding-airbyte/airbyte-protocol/#airbytestream

`source_defined_cursor - If a source supports the INCREMENTAL sync mode, and it sets this field to true, it is responsible for determining internally how it tracks which records in a source are new or updated since the last sync. When set to true, default_cursor_field should also be set.`

**Comment by prateekmukhedkar** [comment]
It seems CDC already sets source_defined_cursor as true, so the ask is to specify a default_cursor_field and a value generated during sync time.

**Comment by evantahler** [comment]
@airbytehq/destinations we'll want to watch this issue

**Comment by prateekmukhedkar** [comment]
Notes from grooming:
1. @akashkulk to verify that this default_cursor_field is not needed for xmin
2. @evantahler we will not be backfilling data. can you confirm if this works?

**Comment by prateekmukhedkar** [comment]
Implementation notes: If we are using a counter, where do we store it? Storing in state makes it tricky

Estimate for Postgres: 3 points

**Comment by evantahler** [comment]
> @evantahler we will not be backfilling data. can you confirm if this works?

I guess that mean users will need to do a full refresh, and this will be a breaking change then.  We will also need every other destination to support this too

**Comment by edgao** [comment]
until we roll out 1s1t, this is a destination-breaking change for sure. SCD tables store the cursor values in the start_at / end_at columns, so if we're changing the cursor then we need to reset. (probably hard reset, if old records don't contain the cursor)

Maybe we can do this nonbreakingly? (e.g. source-postgres checks the configuredcatalog + sees an empty default_cursor_field, and continues using existing behavior. And we have some mechanism to force people to upgrade + refresh schema before 1s1t goes GA)

**Comment by prateekmukhedkar** [comment]
@nguyenaiden  Internal team discussion on how to generate the CDC cursor field https://airbytehq-team.slack.com/archives/C03C4AVJWG4/p1684959706784939?thread_ts=1684880544.887759&cid=C03C4AVJWG4

**Comment by prateekmukhedkar** [comment]
@nguyenaiden will be working on this starting next week (June 12 onwards), scoped to Postgres CDC and for new syncs only. cc @edgao @evantahler

**Comment by evantahler** [comment]
Thanks @prateekmukhedkar! 

That means we can get plan to support Postgres first for the Destinations v2.  Do you have an estimate on the timeline for the other DBs?  We'll need every source to "speak the protocol properly", and have a migration plan, before we can roll out 1-stream-1-table in earnest.

**Comment by prateekmukhedkar** [comment]
1. emit this new field when there is no state - so we only do this for new syncs
2. How does reset state work?

**Comment by nguyenaiden** [comment]
@prateekmukhedkar, with this change, we will continue to not have lsn as a cursor until the customer refresh their source schema as this change happens in the discover phase.

Even if the customer reset their data, they would still not be affected until a whole schema rediscover is carried out.
